gitops/generate.go

```diff
@@ -0,0 +1,204 @@
+)
+
+const (
+	deploymentFileName = "deployment.yaml"
+	serviceFileName    = "service.yaml"
+	routeFileName      = "route.yaml"
+)
+
+// Generate takes in a given Component CR and
+// spits out a deployment, service, and route file to disk
+func generate(fs afero.Fs, outputFolder string, component appstudiov1alpha1.Component) {
+	deployment := generateDeployment(component)
+
+	k := resources.Kustomization{}
+	k.AddResources(deploymentFileName)
+	resources := map[string]interface{}{
+		deploymentFileName: deployment,
+	}
+
+	// If a targetPort was specified, also generate a service and route
+	if component.Spec.TargetPort != 0 {
+		service := generateService(component)
+		route := generateRoute(component)
+		k.AddResources(deploymentFileName, serviceFileName, routeFileName)
+		resources[serviceFileName] = service
+		resources[routeFileName] = route
+	}
+	resources["kustomization.yaml"] = k
+
+	yaml.WriteResources(fs, outputFolder, resources)
+}
+
+func generateDeployment(component appstudiov1alpha1.Component) *appsv1.Deployment {
+	replicas := getReplicas(component)
+
+	labels := map[string]string{
+		"component": component.Name,
+	}
+	deployment := appsv1.Deployment{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      component.Name,
+			Namespace: component.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &v1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: v1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "container-image",
+							Image:           "maven:latest",
+							ImagePullPolicy: corev1.PullAlways,
+							Env:             component.Spec.Env,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Set fields that may have been optionally configured by the component CR
+	if component.Spec.TargetPort != 0 {
+		deployment.Spec.Template.Spec.Containers[0].Ports = []corev1.ContainerPort{
+			{
+				ContainerPort: int32(component.Spec.TargetPort),
+			},
+		}
+		deployment.Spec.Template.Spec.Containers[0].ReadinessProbe = &corev1.Probe{
+			InitialDelaySeconds: 10,
+			PeriodSeconds:       10,
+			Handler: corev1.Handler{
+				TCPSocket: &corev1.TCPSocketAction{
+					Port: intstr.FromInt(component.Spec.TargetPort),
+				},
+			},
+		}
+		deployment.Spec.Template.Spec.Containers[0].LivenessProbe = &corev1.Probe{
+			InitialDelaySeconds: 10,
+			PeriodSeconds:       10,
+			Handler: corev1.Handler{
+				HTTPGet: &corev1.HTTPGetAction{
+					Port: intstr.FromInt(component.Spec.TargetPort),
+					Path: "/",
+				},
+			},
+		}
+	}
+
+	// If the build image is set, set it in the deployment
+	if component.Spec.Build.ContainerImage != "" {
+		deployment.Spec.Template.Spec.Containers[0].Image = component.Spec.Build.ContainerImage
+	}
```

Comment:
I'll remove the if check.

Let's leave the container name as-is for now, since that's the default name that JohnD used in his build+deploy scripts. For a single container deployment, it shouldn't really matter much at this moment.

Reply:
Updated.